Test/tutorial/Simple_compiler/Compiler.dfy

```diff
@@ -0,0 +1,498 @@
+    : DafnyAST.Stmt
+    reads *
+  {
+    if c is CSharpAST.Print then
+      var c := c as CSharpAST.Print;
+      DafnyAST.Print(translateExpr(c.e))
+    else
+      translateStmt(c)
+  }
+
+  function method {:verify false} translateProg(c: CSharpAST.Prog)
+    : DafnyAST.Stmt
+    reads *
+  {
+    CSharpUtils.ListUtils.FoldR(
+      (c, ds) => DafnyAST.Seq(translateStmt(c), ds),
+      DafnyAST.Skip,
+      c.s
+    )
+  }
+}
+
+/// Pretty-printing Dafny's outputs
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///
+/// In general passing complex types between Dafny and C# requires some efforts (e.g. the modeling above), so for the final part of this pipeline (returning the data to C#) we'll use a very simple type: a list of strings (one per instruction).  Conveniently, the `ToString` method of Dafny strings provides a trivial conversion to C# strings, so we leave that conversion to the C# side.
+///
+/// Alternatively, we could have chosen to expose the stack machine types to C# directly and do the pretty-printing from there.
+
+module PrettyPrint {
+  import opened LinkedList
+  import opened StackMachine
+
+  function method prettyPrintNum(n: int, zero: string) : string
+    decreases n < 0, if n < 0 then -n else n
+  {
+    if n == 0 then zero
+    else if n < 0 then prettyPrintNum(-n, zero)
+    else if n < 10 then ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"][n]
+    else if n >= 10 then prettyPrintNum(n / 10, "") + prettyPrintNum(n % 10, "0")
+    else assert false; prettyPrintNum(n, zero)
+  }
+
+  function method prettyPrintInstr(instr: Instr) : string {
+    match instr {
+      case PushConst(n) => "PushConst(" + prettyPrintNum(n, "0") + ")"
+      case PopAdd => "PopAdd"
+      case PopSub => "PopSub"
+      case Print => "Print"
+    }
+  }
+
+  function method prettyPrint(p: Prog) : List<string> {
+    match p {
+      case Nil => Nil
+      case Cons(instr, p) => Cons(prettyPrintInstr(instr), prettyPrint(p))
+    }
+  }
+}
+
+/// 5. Exposing Dafny to C#
+/// ~~~~~~~~~~~~~~~~~~~~~~~
+///
+/// Finally, we define a single Dafny function that serves as our interface to C#.  It takes a C# AST, translates it to Dafny, runs rewriting passes, compiles it to a stack machine program, pretty-prints it to a string, and returns it to C#.
+///
+/// Note that the `Interop` module below has an `extern` annotation but no `{:compile false}` annotation: this is because we want to implement it in Dafny, but give it a predictable name in the generated C# code.
+
+module {:extern "SimpleCompiler"} Interop {
+  import LinkedList
+  import CSharpAST
+  import DafnyAST
+  import Rewriter
+  import StackMachine
+  import Translator
+  import Compiler
+  import PrettyPrint
+  import CSharpUtils
+  import Generics = System.Collections.Generic
+
+  class DafnyCompiler {
+    static method Compile(dAST: DafnyAST.Stmt) returns (dSM: StackMachine.Prog)
+      ensures DafnyAST.interpStmt(dAST) == StackMachine.interpProg(dSM)
```

Comment:
Cool postcondition!! Would it be more a more complete example if these programs could also handle input? I was looking for the `forall input ::` but I see it doesn't apply to this language.

Reply:
Good point, I revamped the example to allow variables, so you can simulate input by running the program in a nonempty context.  Input in the sense of I/O would be harder to model (it requires nondeterminism)

Reply:
Could you still change `CompileCorrect` to

```
  lemma compileCorrect(s: DafnyAST.Stmt)
    ensures forall input: DafnyAST.Context :: interpProg(compileStmt(s), input) == DafnyAST.interpStmt(s, input)
```

?

Reply:
Sure thing, though of course that requires changing the types of the methods on both sides; done in the latest commit

Reply:
Thanks :-)